File: .devcontainer/skills/quint-loop/references/examples/crash-durability/replay.test.ts
// Replay the Quint counter-example (session.itf.json, produced by session.neg.qnt)
// against a minimal re-implementation of two persistence patterns:
//
//   Buggy:  truncate the canonical path, then write in-place.  A crash
//           between truncate and write leaves disk empty — data loss.
//   Fixed:  write to a sibling tmp file, then rename.  The canonical path
//           is never in an intermediate state.
//
// For Layer 2 idioms (`vi.mock('node:fs/promises', ...)` to force a
// mid-write crash against real code), see SKILL.md step 7.

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { describe, expect, it } from "vitest";

const here = dirname(fileURLToPath(import.meta.url));

type ItfState = {
  memVersion: { "#bigint": string };
  diskVersion: { "#bigint": string };
  writeInProgress: boolean;
  committedOnce: boolean;
};
type Itf = { states: ItfState[] };

type Action =
  | { kind: "writeNew" }
  | { kind: "beginSave" }
  | { kind: "finishSave" }
  | { kind: "crash" };

function num(b: { "#bigint": string }): number {
  return Number(b["#bigint"]);
}

// Derive actions from state diffs. Quint ITFs don't record the action name,
// so we infer from which vars changed.
function deriveActions(states: ItfState[]): Action[] {
  const out: Action[] = [];
  for (let i = 1; i < states.length; i++) {
    const p = states[i - 1];
    const c = states[i];
    const memDelta = num(c.memVersion) - num(p.memVersion);
    const wipOn = !p.writeInProgress && c.writeInProgress;
    const wipOff = p.writeInProgress && !c.writeInProgress;
    const diskRose = num(c.diskVersion) > num(p.diskVersion);
    if (memDelta > 0) out.push({ kind: "writeNew" });
    else if (wipOn) out.push({ kind: "beginSave" });
    else if (wipOff && diskRose) out.push({ kind: "finishSave" });
    else if (wipOff) out.push({ kind: "crash" });
  }
  return out;
}

// --- Buggy mirror: in-place truncate-then-write. ---
export class BuggyStore {
  diskVersion = 0; // 0 = empty
  memVersion = 0;
  writeInProgress = false;
  committedOnce = false;

  writeNew() {
    this.memVersion++;
  }
  beginSave() {
    this.writeInProgress = true;
    this.diskVersion = 0; // the bug: truncate in place
  }
  finishSave() {
    this.diskVersion = this.memVersion;
    this.writeInProgress = false;
    this.committedOnce = true;
  }
  crash() {
    this.writeInProgress = false;
    // disk unchanged — stays at whatever beginSave left
  }
  load(): number {
    return this.diskVersion;
  }
}

// --- Fixed mirror: tmp+rename. ---
export class FixedStore {
  diskVersion = 0;
  memVersion = 0;
  private tmpVersion: number | null = null;
  committedOnce = false;

  writeNew() {
    this.memVersion++;
  }
  beginSave() {
    this.tmpVersion = this.memVersion; // write to tmp; disk untouched
  }
  finishSave() {
    if (this.tmpVersion === null) throw new Error("no tmp staged");
    this.diskVersion = this.tmpVersion; // atomic rename
    this.tmpVersion = null;
    this.committedOnce = true;
  }
  crash() {
    this.tmpVersion = null; // abandon tmp
  }
  load(): number {
    return this.diskVersion;
  }
  get writeInProgress(): boolean {
    return this.tmpVersion !== null;
  }
}

describe("crash-durability (replay of Quint trace)", () => {
  const trace: Itf = JSON.parse(
    readFileSync(join(here, "session.itf.json"), "utf-8"),
  );
  const actions = deriveActions(trace.states);

  it("buggy mirror loses committed data on the pinned trace", () => {
    const s = new BuggyStore();
    for (const a of actions) {
      if (a.kind === "writeNew") s.writeNew();
      else if (a.kind === "beginSave") s.beginSave();
      else if (a.kind === "finishSave") s.finishSave();
      else if (a.kind === "crash") s.crash();
    }
    // Trace reaches state 6: committedOnce=true, diskVersion=0. Data loss.
    expect(s.committedOnce).toBe(true);
    expect(s.load()).toBe(0);
  });

  it("fixed mirror preserves committed data under the same trace", () => {
    const s = new FixedStore();
    for (const a of actions) {
      if (a.kind === "writeNew") s.writeNew();
      else if (a.kind === "beginSave") s.beginSave();
      else if (a.kind === "finishSave") s.finishSave();
      else if (a.kind === "crash") s.crash();
    }
    // committedOnce implies load() > 0 — the invariant holds.
    if (s.committedOnce) expect(s.load()).toBeGreaterThan(0);
  });
});

File: .devcontainer/skills/quint-loop/references/examples/crash-durability/replay_test.py
"""Replay the Quint counter-example (session.itf.json, produced by session.neg.qnt)
against a minimal re-implementation of two persistence patterns.

  Buggy:  truncate the canonical path, then write in-place.  A crash
          between truncate and write leaves disk empty — data loss.
  Fixed:  write to a sibling tmp file, then rename.  The canonical path
          is never in an intermediate state.

For Layer 2 idioms (mocking `os.write` or `builtins.open` to simulate a
mid-write crash against real code), see SKILL.md step 7.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

import pytest


HERE = Path(__file__).parent


@dataclass
class Action:
    kind: str  # "writeNew" | "beginSave" | "finishSave" | "crash"


def _num(v: dict) -> int:
    return int(v["#bigint"])


def derive_actions(states: list[dict]) -> list[Action]:
    out: list[Action] = []
    for i in range(1, len(states)):
        p, c = states[i - 1], states[i]
        mem_delta = _num(c["memVersion"]) - _num(p["memVersion"])
        wip_on = not p["writeInProgress"] and c["writeInProgress"]
        wip_off = p["writeInProgress"] and not c["writeInProgress"]
        disk_rose = _num(c["diskVersion"]) > _num(p["diskVersion"])
        if mem_delta > 0:
            out.append(Action("writeNew"))
        elif wip_on:
            out.append(Action("beginSave"))
        elif wip_off and disk_rose:
            out.append(Action("finishSave"))
        elif wip_off:
            out.append(Action("crash"))
    return out


# --- Buggy mirror: in-place truncate-then-write. ---
class BuggyStore:
    def __init__(self) -> None:
        self.disk_version = 0
        self.mem_version = 0
        self.write_in_progress = False
        self.committed_once = False

    def write_new(self) -> None:
        self.mem_version += 1

    def begin_save(self) -> None:
        self.write_in_progress = True
        self.disk_version = 0  # the bug: truncate in place

    def finish_save(self) -> None:
        self.disk_version = self.mem_version
        self.write_in_progress = False
        self.committed_once = True

    def crash(self) -> None:
        self.write_in_progress = False
        # disk unchanged

    def load(self) -> int:
        return self.disk_version


# --- Fixed mirror: tmp+rename. ---
class FixedStore:
    def __init__(self) -> None:
        self.disk_version = 0
        self.mem_version = 0
        self._tmp_version: int | None = None
        self.committed_once = False

    def write_new(self) -> None:
        self.mem_version += 1

    def begin_save(self) -> None:
        self._tmp_version = self.mem_version  # stage to tmp, disk untouched

    def finish_save(self) -> None:
        if self._tmp_version is None:
            raise RuntimeError("no tmp staged")
        self.disk_version = self._tmp_version  # atomic rename
        self._tmp_version = None
        self.committed_once = True

    def crash(self) -> None:
        self._tmp_version = None  # abandon tmp

    def load(self) -> int:
        return self.disk_version

    @property
    def write_in_progress(self) -> bool:
        return self._tmp_version is not None


def _apply(store, action: Action) -> None:
    dispatch = {
        "writeNew": store.write_new,
        "beginSave": store.begin_save,
        "finishSave": store.finish_save,
        "crash": store.crash,
    }
    dispatch[action.kind]()


@pytest.fixture(scope="module")
def actions() -> list[Action]:
    trace = json.loads((HERE / "session.itf.json").read_text())
    return derive_actions(trace["states"])


def test_buggy_store_loses_committed_data(actions: list[Action]) -> None:
    s = BuggyStore()
    for a in actions:
        _apply(s, a)
    # Trace reaches a state where committedOnce=True but diskVersion=0.
    assert s.committed_once is True
    assert s.load() == 0


def test_fixed_store_preserves_committed_data(actions: list[Action]) -> None:
    s = FixedStore()
    for a in actions:
        _apply(s, a)
    if s.committed_once:
        assert s.load() > 0
